flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -109,12 +110,12 @@ public DynamicTableSinkSpec getTableSinkSpec() {
 
     @SuppressWarnings("unchecked")
     protected Transformation<Object> createSinkTransformation(
-            StreamExecutionEnvironment env,
-            TableConfig tableConfig,
+            PlannerBase planner,
             Transformation<RowData> inputTransform,
             int rowtimeFieldIndex,
             boolean upsertMaterialize) {
-        final DynamicTableSink tableSink = tableSinkSpec.getTableSink();
+        final DynamicTableSink tableSink = tableSinkSpec.getTableSink(planner);
+        final ChangelogMode inputChangelogMode = tableSink.getChangelogMode(changelogMode);
```

Comment:
sorry, but I still fine this confusing. Shouldn't this variable be called `changelogMode` and the member `inputChangelogMode`